TESTS/psa/crypto_init/main.cpp

```diff
@@ -0,0 +1,104 @@
+/*
+* Copyright (c) 2018 ARM Limited. All rights reserved.
+*
+* SPDX-License-Identifier: Apache-2.0
+*
+* Licensed under the Apache License, Version 2.0 (the License); you may
+* not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an AS IS BASIS, WITHOUT
+* WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include "greentea-client/test_env.h"
+#include "unity/unity.h"
+#include "utest/utest.h"
+#include "crypto.h"
+#include "entropy.h"
+#include "entropy_poll.h"
+
+#if (!defined(TARGET_PSA))
+#error [NOT_SUPPORTED] PSA entropy injection tests can run only on PSA-enabled targets.
+#endif // TARGET_PSA
+
+#define TEST_RANDOM_SIZE 64
+
+#if !defined(MAX)
+#define MAX(a,b) (((a)>(b))?(a):(b))
+#endif
+
+/* Calculating the minimum allowed entropy size in bytes */
+#define MBEDTLS_PSA_INJECT_ENTROPY_MIN_SIZE \
+            MAX(MBEDTLS_ENTROPY_MIN_PLATFORM, MBEDTLS_ENTROPY_BLOCK_SIZE)
+
+using namespace utest::v1;
+
+utest::v1::status_t greentea_test_setup(const size_t number_of_cases)
+{
+#ifndef NO_GREENTEA
+    GREENTEA_SETUP(60, "default_auto");
+#endif
+    return greentea_test_setup_handler(number_of_cases);
+}
+
+static void check_multi_crypto_init_deinit()
+{
+    uint8_t output[TEST_RANDOM_SIZE] = {0};
+    uint8_t seed[MBEDTLS_PSA_INJECT_ENTROPY_MIN_SIZE] = {0};
+    /* inject some a seed for test*/
+    for(int i; i < MBEDTLS_PSA_INJECT_ENTROPY_MIN_SIZE; ++i)
+        seed[i] = i;
+    /* don't really care if this succeed this is just to make crypto init pass*/
+    mbedtls_psa_inject_entropy( seed, MBEDTLS_PSA_INJECT_ENTROPY_MIN_SIZE );
+    psa_status_t status = psa_crypto_init();
+    TEST_ASSERT_EQUAL(PSA_SUCCESS, status);
+    status = psa_crypto_init();
+    TEST_ASSERT_EQUAL(PSA_SUCCESS, status);
+    status = psa_generate_random(output, sizeof(output));
+    TEST_ASSERT_NOT_EQUAL(PSA_ERROR_BAD_STATE, status);
+    mbedtls_psa_crypto_free();
+    status = psa_generate_random(output, sizeof(output));
+    TEST_ASSERT_NOT_EQUAL(PSA_ERROR_BAD_STATE, status);
```

Comment:
Major: You've just closed the connection to the crypto library, so `psa_generate_random` should return `BAD_STATE`.

I see that you've implemented a connection counter in the SPE. It makes sense, but this violates the documented semantics of `psa_crypto_init` and `mbedtls_psa_crypto_free`: multiple calls to `psa_crypto_init` are supposed to be equivalent to a single call and a single call to `mbedtls_psa_crypto_free` is supposed to always free all resources associated with the current client.

Should the semantics change? My reasoning for the current semantics is that you should be able to call `mbedtls_psa_crypto_free` in your exit code regardless of how many times `mbedtls_psa_crypto_free` has been called. If you think the current semantics is wrong and the semantics in this PR is better, please submit PSA crypto API design feedback via internal or [external](https://github.com/ARMmbed/mbed-crypto#user-content-feedback-welcome) channels that explains your reasoning.

This deviation needs to be documented in the known issues for this release.